fluid/DeepQNetwork/train.py

```diff
@@ -0,0 +1,188 @@
+#-*- coding: utf-8 -*-
+
+from DQN_agent import DQNModel
+from DoubleDQN_agent import DoubleDQNModel
+from DuelingDQN_agent import DuelingDQNModel
+from atari import AtariPlayer
+import paddle.fluid as fluid
+import gym
+import argparse
+import cv2
+from tqdm import tqdm
+from expreplay import ReplayMemory, Experience
+import numpy as np
+import os
+
+from datetime import datetime
+from atari_wrapper import FrameStack, MapState, FireResetEnv, LimitLength
+from collections import deque
+
+HIST_LEN = 4
+UPDATE_FREQ = 4
+
+#MEMORY_WARMUP_SIZE = 2000
+MEMORY_SIZE = 1e6
+MEMORY_WARMUP_SIZE = MEMORY_SIZE // 20
+IMAGE_SIZE = (84, 84)
+FRAME_HISTORY = 4
+ACTION_REPEAT = 4  # aka FRAME_SKIP
+UPDATE_FREQ = 4
+
+
+def run_train_episode(agent, env, exp):
+    total_reward = 0
+    state = env.reset()
+    step = 0
+    while True:
+        step += 1
+        history = exp.recent_state()
+        history.append(state)
+        history = np.stack(history, axis=0)
+        action = agent.act(history, train_or_test='train')
```

Comment:
Here you use history to do prediction. But in eval_agent() and play.py you only use one frame? Why are they not consistent?

Reply:
Not only use one frame here.The env used in eval_agent() is wrapped by  **FrameStack** wrapper, which restores the recent frames in environment.And in training process, recent frames were taken care of by replay memory.
https://github.com/zenghsh3/models/blob/3bf1d4b2b48e4239f4b420c8f3592d7d47e1bb33/fluid/DeepQNetwork/atari_wrapper.py#L26